File: 225_implement_stack_using_queues.java
// Use two queues, always use q1 as main queue, and q2 as buffer
class MyStack {
    // Push element x onto stack.
    public void push(int x) {
        q2.add(x);
        while(!q1.isEmpty()) {
            q2.add(q1.element());
            q1.remove();
        }
        LinkedList<Integer> t = q1;
        q1 = q2;
        q2 = t;
    }

    // Removes the element on top of the stack.
    public void pop() {
        q1.remove();
    }

    // Get the top element.
    public int top() {
        return q1.element();
    }

    // Return whether the stack is empty.
    public boolean empty() {
        return q1.isEmpty();
    }
    
    private LinkedList<Integer> q1 = new LinkedList<Integer>();
    private LinkedList<Integer> q2 = new LinkedList<Integer>();
}

File: 225_implement_stack_using_queues.cc
// Use to queues, when adding an element, put into the empty queue, and
// migrate all elements from the other queue to this queue.

class Stack {
public:
    // Push element x onto stack.
    void push(int x) {
        if(q1.empty()) {
            q1.push_back(x);
            while(!q2.empty()) {
                q1.push_back(q2.front());
                q2.pop_front();
            }
        } else {
            q2.push_back(x);
            while(!q1.empty()) {
                q2.push_back(q1.front());
                q1.pop_front();
            }
        }
    }

    // Removes the element on top of the stack.
    void pop() {
        if (!q1.empty()) q1.pop_front();
        else q2.pop_front();
    }

    // Get the top element.
    int top() {
        if (!q1.empty()) return q1.front();
        else return q2.front();
    }

    // Return whether the stack is empty.
    bool empty() {
        return (q1.empty() && q2.empty());
    }

private:
    deque<int> q1;
    deque<int> q2;
};
